plugins/spidmx/SPIDMXThread.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * SPIDMXThread.cpp
+ * This thread runs while one or more ports are registered. It simultaneously
+ * reads / writes SPI data and then calls the parser. This is repeated forever.
+ * Copyright (C) 2017 Florian Edelmann
+ */
+
+#include <string>
+#include "ola/Logging.h"
+#include "plugins/spidmx/SPIDMXWidget.h"
+#include "plugins/spidmx/SPIDMXThread.h"
+#include "plugins/spidmx/SPIDMXParser.h"
+
+namespace ola {
+namespace plugin {
+namespace spidmx {
+
+SPIDMXThread::SPIDMXThread(SPIDMXWidget *widget, unsigned int blocklength)
+  : m_widget(widget),
+    m_blocklength(blocklength),
+    m_term(false),
+    m_registered_ports(0),
+    m_spi_rx_buffer(blocklength),
+    m_spi_tx_buffer(blocklength) {
+}
+
+SPIDMXThread::~SPIDMXThread() {
+  Stop();
+}
+
+/**
+ * This thread does only have to run if ports using it are patched to a
+ * universe. Thus, they must register and unregister to notify this thread.
+ */
+void SPIDMXThread::RegisterPort() {
+  m_registered_ports++;
+
+  if (m_registered_ports >= 1) {
+    Start();
+  }
+}
+void SPIDMXThread::UnregisterPort() {
+  m_registered_ports--;
+
+  if (m_registered_ports <= 0) {
+    Stop();
+  }
+}
+
+
+/**
+ * Stop this thread
+ */
+bool SPIDMXThread::Stop() {
+  {
+    ola::thread::MutexLocker locker(&m_term_mutex);
+    m_term = true;
+  }
+  return Join();
+}
+
+
+/**
+ * Copy a DmxBuffer to the output thread
+ */
+bool SPIDMXThread::WriteDMX(const DmxBuffer &buffer) {
+  ola::thread::MutexLocker locker(&m_buffer_mutex);
+  m_dmx_tx_buffer.Set(buffer);
+  return true;
+}
+
+/**
+  * @brief Get DMX Buffer
+  * @returns DmxBuffer with current input values.
+  */
+const DmxBuffer &SPIDMXThread::GetDmxInBuffer() const {
+  return m_dmx_rx_buffer;
+}
+
+
+/**
+  * @brief Set the callback to be called when the receive buffer is updated.
+  * @param callback The callback to call.
+  */
+bool SPIDMXThread::SetReceiveCallback(Callback0<void> *callback) {
+  if (!callback) {
```

Comment:
Shouldn't this either be m_receive_callback, or the line below do a reset to NULL?

Reply:
Oh right, thanks!

Reply:
No worries, done.